core/jobs/batch_jobs/manual_voice_artist_names_job.py

```diff
@@ -0,0 +1,504 @@
+        self.total_commit_log_models = len(exp_commit_log_models)
+        self.processed_commit_log_models += 1
+
+        if self.processed_commit_log_models == self.total_commit_log_models:
+            voice_artist_metadata_models = self._convert_to_models(
+                voice_artist_id_to_metadata_mapping)
+
+            for voice_artist_metadata_model in voice_artist_metadata_models:
+                yield voice_artist_metadata_model
+
+
+class CreateVoiceArtistMetadataModelsFromExplorationsJob(base_jobs.JobBase):
+    """Jobs used for fetching and saving voice artist names from curated
+    exploration models.
+    """
+
+    DATASTORE_UPDATES_ALLOWED = True
+    voice_artist_metadata_mapping: Dict[
+        str, voiceover_models.VoiceoversAndContentsMappingType] = {}
+
+    @staticmethod
+    def _delete_voice_artist_model_for_old_commits(
+        language_code: str,
+        exploration_id: str,
+        content_id: str,
+        voice_artist_id_to_metadata_mapping: Dict[
+            str, voiceover_models.VoiceoversAndContentsMappingType]
+    ) -> Dict[str, voiceover_models.VoiceoversAndContentsMappingType]:
+        """Remove the content ID association from the metadata of the previous
+        voice artist when a new voiceover is recorded for the same content
+        by another voice artist, necessitating the deletion of the
+        previous record.
+
+        Args:
+            language_code: str. The language code in which the new voiceover
+                is added.
+            exploration_id: str. The exploration ID in which the new voiceover
+                is added.
+            content_id: str. The content ID in which the new voiceover is added.
+            voice_artist_id_to_metadata_mapping:
+                dict(str, VoiceoversAndContentsMappingType). A dictionary
+                mapping voice artist IDs as keys and nested dicts as values.
+                Each nested dict refers to the voiceovers_and_contents_mapping
+                field in the VoiceArtistMetadataModel.
+
+        Returns:
+            dict(str, VoiceoversAndContentsMappingType). A dictionary
+            mapping voice artist IDs as keys and nested dicts as values.
+            Each nested dict refers to the voiceovers_and_contents_mapping
+            field in the VoiceArtistMetadataModel.
+        """
+        voice_artist_id_to_metadata_mapping_iterable: Dict[
+            str, voiceover_models.VoiceoversAndContentsMappingType] = (
+                copy.deepcopy(voice_artist_id_to_metadata_mapping))
+
+        for voice_artist_id, voiceovers_and_contents_mapping in (
+                voice_artist_id_to_metadata_mapping_iterable.items()):
+            language_mapping = voiceovers_and_contents_mapping.get(
+                language_code, {})
+            # Here we use cast because we are narrowing down the type from
+            # Union[str, Dict[str, List[str]], Dict[str, List[VoiceoverDict]]
```

Comment:
I don't quite understand this. Why is the type ambiguous in the first place?

Reply:
Basically voiceovers_and_contents_mapping can be any of str / dict(str, list(str)) / dict(str, list(voiceover)) types. So if we try to get content_ids (which is str) mypy is raising error expected Union[....] similar in case of voiceovers. Thus I have narrowed down the scope.

Reply:
But why is voiceovers_and_contents_mapping allowed to have all those types? That is too broad, and it doesn't make sense for that to be a string, for example.

Reply:
Removed